File: platform/android/MapLibreAndroid/src/webgpu/java/org/maplibre/android/maps/renderer/MapRendererFactory.java
package org.maplibre.android.maps.renderer;

import android.content.Context;
import android.view.Surface;
import android.view.TextureView;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;

import org.maplibre.android.maps.renderer.surfaceview.MapLibreWebGPUSurfaceView;
import org.maplibre.android.maps.renderer.surfaceview.SurfaceViewMapRenderer;
import org.maplibre.android.maps.renderer.surfaceview.WebGPUSurfaceViewMapRenderer;
import org.maplibre.android.maps.renderer.textureview.TextureViewMapRenderer;
import org.maplibre.android.maps.renderer.textureview.WebGPUTextureViewRenderThread;

@Keep
public class MapRendererFactory {
  public static TextureViewMapRenderer newTextureViewMapRenderer(@NonNull Context context, TextureView textureView,
                                                                 String localFontFamily, boolean translucentSurface,
                                                                 Runnable initCallback) {

    TextureViewMapRenderer mapRenderer = new TextureViewMapRenderer(context, textureView,
            localFontFamily, translucentSurface) {
      @Override
      protected void onSurfaceCreated(Surface surface) {
        initCallback.run();
        super.onSurfaceCreated(surface);
      }
    };

    mapRenderer.setRenderThread(new WebGPUTextureViewRenderThread(textureView, mapRenderer));
    return mapRenderer;
  }

  public static SurfaceViewMapRenderer newSurfaceViewMapRenderer(@NonNull Context context, String localFontFamily,
                                                                 boolean renderSurfaceOnTop, Runnable initCallback) {

    MapLibreWebGPUSurfaceView surfaceView = new MapLibreWebGPUSurfaceView(context);
    surfaceView.setZOrderMediaOverlay(renderSurfaceOnTop);

    return new WebGPUSurfaceViewMapRenderer(context, surfaceView, localFontFamily) {
      @Override
      public void onSurfaceCreated(Surface surface) {
        initCallback.run();
        super.onSurfaceCreated(surface);
      }
    };
  }
}

File: platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/maps/renderer/MapRendererFactory.kt
package org.maplibre.android.maps.renderer

import android.content.Context
import android.view.Surface
import android.view.TextureView
import androidx.annotation.Keep
import org.maplibre.android.maps.renderer.surfaceview.SurfaceViewMapRenderer
import org.maplibre.android.maps.renderer.textureview.TextureViewMapRenderer

/**
 * Shared factory used by MapRenderer.create(). The shape of renderer construction
 * (anonymous subclass wiring of initCallback, render-thread attachment) lives here;
 * the backend-specific concrete-type instantiation is delegated to the
 * flavor-provided `RendererStrategy`.
 */
@Keep
object MapRendererFactory {
    @JvmStatic
    fun newTextureViewMapRenderer(
        context: Context,
        textureView: TextureView,
        localFontFamily: String?,
        translucentSurface: Boolean,
        initCallback: Runnable,
    ): TextureViewMapRenderer {
        val mapRenderer =
            object : TextureViewMapRenderer(context, textureView, localFontFamily, translucentSurface) {
                override fun onSurfaceCreated(surface: Surface?) {
                    initCallback.run()
                    super.onSurfaceCreated(surface)
                }
            }
        RendererStrategy.attachTextureRenderThread(textureView, mapRenderer)
        return mapRenderer
    }

    @JvmStatic
    fun newSurfaceViewMapRenderer(
        context: Context,
        localFontFamily: String?,
        renderSurfaceOnTop: Boolean,
        initCallback: Runnable,
    ): SurfaceViewMapRenderer =
        RendererStrategy.createSurfaceViewRenderer(
            context,
            localFontFamily,
            renderSurfaceOnTop,
            initCallback,
        )
}
